Add tests, api, models, auth module and migrations by alembic

DaloreVacAtion opened the pull request:
Добавлены апишки, модели, авторизация, миграция. и тесты, на которые потрачено больше всего времени

Commit:
Add tests, api, models, auth module and migrations by alembic

maximovd commented on api/router.py:
```diff
@@ -0,0 +1,42 @@
+"""
+*********************
+Модуль app->api->router.py
```
А какой резон писать такие комментарии ?

DaloreVacAtion replied:
Сначала думал, что буду помечать, чтобы понятнее было, потом забил в оставшемся коде. Удалю.

maximovd commented on api/router.py:
```diff
@@ -0,0 +1,42 @@
+Создание APIRouter для главной ветки в FastAPI
+"""
+
+# fast api users
```
Опять же, не очень понятно, зачем тут этот коммент

DaloreVacAtion replied:
удалил

maximovd commented on api/v1/crud.py:
```diff
@@ -0,0 +1,28 @@
+
+
+async def get_user_by_id(user_id: int, session: AsyncSession):
+    query = text(f'select * from public.user where public.user.id = {user_id}')
```
Это не должно лежать в `api`. Это по сути у тебя сервисы. Поэтому нужен дополнительный слой, который будет содержать сервисы. Что-то типа `app/services/user.py`

DaloreVacAtion replied:
Перекинул

maximovd commented on api/v1/router.py:
```diff
@@ -0,0 +1,84 @@
+        username: Annotated[str, Query(title='Никнейм пользователя')],
+        session: AsyncSession = Depends(get_async_session),
+) -> UserRead:
+    # query = select(User).where(User.username == username)  # Почему так не работат?
```
Зачем тут этот некрокод ?

DaloreVacAtion replied:
Тут у меня были вопросы, ибо я сделал по примеру из документации/интернета, но почему-то он выдавал 404, что такого пользователя нет, хотя при "сыром" запросе, который указан ниже, пользователь выводится корректно. Насколько я знаю, сырые запросы могут быть не безопасны, в случае, если плохая защита и могут произойти SQL инъекции

maximovd commented on api/v1/router.py:
```diff
@@ -0,0 +1,84 @@
+        session: AsyncSession = Depends(get_async_session),
+) -> UserRead:
+    # query = select(User).where(User.username == username)  # Почему так не работат?
+    query = text(f'select * from public.user where public.user.username = \'{username}\'')
```
А почему тут не убрал в сервис ? Не стоит оставлять бизнес-логику в контроллерах. Это стоит вынести в `services`

DaloreVacAtion replied:
Понял, сделаю

DaloreVacAtion replied:
Вынес

maximovd commented on api/v1/router.py:
```diff
@@ -0,0 +1,84 @@
+        user_id: Annotated[int, Path(title='ID пользователя', ge=0)],
+        user_in: UserUpdate,
+        session: AsyncSession = Depends(get_async_session),
+):
```
Где выходной тип ?

DaloreVacAtion replied:
Добавлено

maximovd commented on api/v1/router.py:
```diff
@@ -0,0 +1,84 @@
+        user_in: UserUpdate,
+        session: AsyncSession = Depends(get_async_session),
+):
+    """Обновить пользователя"""
```
По правилам, в конце доки надо ставить точку. Можно почитать вот тут https://peps.python.org/pep-0257/

DaloreVacAtion replied:
Спасибо, добавлю

maximovd commented on api/v1/router.py:
```diff
@@ -0,0 +1,84 @@
+            detail=f"Пользователя с id {user_id} не существет",
+        )
+    await crud.update_user(user_id, user_in, session)
+    # Как вернуть нормального пользоваетеля?
```
Что это значит ?

DaloreVacAtion replied:
Уточнил в вопросах, прикреплю позже

maximovd commented on api/v1/router.py:
```diff
@@ -0,0 +1,84 @@
+    await crud.update_user(user_id, user_in, session)
+    # Как вернуть нормального пользоваетеля?
+    await session.refresh(user)
+    return {'message': f'Successfully updated user'}
```
Лучше возвращать схему, а не просто словарь

DaloreVacAtion replied:
Это тоже в вопросах

maximovd commented on auth/base_config.py:
```diff
@@ -0,0 +1,18 @@
+
+from config import SECRET_KEY
+
+cookie_transport = CookieTransport(cookie_name='IDP', cookie_max_age=3600)
```
А зачем тебе тут куки ?

DaloreVacAtion replied:
В условиях сказано, что все запросы должны содержать Token, поэтому я решил использовать самый простой, JWT. Я погуглил и узнал, что для того, чтобы передать JWT Token надо куки прописать. Не знаю, насколько это крректно

maximovd commented on auth/custon_authenticator.py:
```diff
@@ -0,0 +1,5 @@
+from fastapi_users.authentication import Authenticator
+
+
+class CustomAuthenticator(Authenticator):
```
Зачем ?

DaloreVacAtion replied:
Хотел сделать свой кастомный аутентификатор, но потом подумал, что в условиях задачи этого нет, а значит не нужно лишних действий, а убрать забыл

maximovd commented on auth/manager.py:
```diff
@@ -0,0 +1,22 @@
+    reset_password_token_secret = SECRET
+    verification_token_secret = SECRET
+
+    async def on_after_register(self, user: User, request: Optional[Request] = None):
```
Какой ты используешь python ? В python3.10 хинты лучше писать вот так: 
```python
request: Request | None = None
```

DaloreVacAtion replied:
Заменил

maximovd commented on auth/manager.py:
```diff
@@ -0,0 +1,22 @@
+    verification_token_secret = SECRET
+
+    async def on_after_register(self, user: User, request: Optional[Request] = None):
+        print(f"User {user.id} has registered.")
```
Не очень понял две вещи:

1. Зачем тут `print` ? Принты в коде вообще не должны попадаться, это потенциальная точка слива данных
2. Зачем вообще нужен этот метод ?

DaloreVacAtion replied:
Делал просто по примеру, поэтому на автомате добавил метод. По идее, его можно использовать после регистрации, например, чтобы отправить e-mail с подтверждением, а так да, в данном контексте он бесполезный

maximovd commented on auth/models.py:
```diff
@@ -0,0 +1,30 @@
+    """
+    Модель пользователя для ДБ
+    """
+    # поля нашего пользвоателя
```
А какая разница ? Зачем нам их как-то отделять ?

DaloreVacAtion replied:
Просто решил пометить

maximovd commented on config.py:
```diff
@@ -0,0 +1,20 @@
+
+load_dotenv()
+
+DB_USER = os.environ.get('DB_USER')
```
https://fastapi.tiangolo.com/advanced/settings/#install-pydantic-settings
почитай вот тут, ИМХО, но такой вариант приятней глазу и удобнее за счет возможности валидации входящих перменных окружения

DaloreVacAtion replied:
Добавил Сеттингсы

maximovd commented on core/logging.py:
```diff
@@ -0,0 +1,47 @@
+import abc
+import datetime
+import logging
+from loguru import logger
```
`loguru` все-таки не очень хорош, давай попробуем заменить его на https://www.structlog.org/en/stable/

DaloreVacAtion replied:
Ок

maximovd commented on core/resources.py:
```diff
@@ -0,0 +1,8 @@
+
+class ServiceResult(BaseModel):
+    ok: bool
+    message: Optional[str]
```
Проверь везде по коду хинты: `Optional[str]` -> `str | None`

DaloreVacAtion replied:
Заменил

maximovd commented on core/settings.py:
```diff
@@ -0,0 +1,18 @@
+
+
+# ####### Main application settings #########
+ENV = config.get("ENV", cast=str, default="production")
```
Так а зачем тебе отдельно `settings` и отдельно `config` ? Можно все объединить в класс Pydantic-Config

DaloreVacAtion replied:
Объеденил

maximovd commented on db/database.py:
```diff
@@ -0,0 +1,21 @@
+# тут получается подключаемся к БД
```
Зачем такие комменты ?

DaloreVacAtion replied:
Удалил, добавлял больше для себя

maximovd commented on db/database.py:
```diff
@@ -0,0 +1,21 @@
+
+from config import DB_USER, DB_PASS, DB_HOST, DB_PORT, DB_NAME
+
+DATABASE_URL = f'postgresql+asyncpg://{DB_USER}:{DB_PASS}@{DB_HOST}:{DB_PORT}/{DB_NAME}'
```
Когда добавишь `pydantic-settings`, можно билд этого URI убрать в валидатор

DaloreVacAtion replied:
УБрал

maximovd commented on main.py:
```diff
@@ -0,0 +1,24 @@
+    debug=settings.DEBUG
+)
+
+# logger
```
Тут из контекста понятно, что это логгер, комментарий не обязателен

DaloreVacAtion replied:
Убрал

maximovd commented on main.py:
```diff
@@ -0,0 +1,24 @@
+# logger
+app.logger = logger
+
+# Мидлвари
```
См. коммент выше

DaloreVacAtion replied:
Убрал

maximovd commented on test_main.http:
```diff
@@ -0,0 +1,11 @@
+# Test your FastAPI endpoints
+
+GET http://127.0.0.1:8000/
```
А оно нам точно надо ? Мы прям пользоваться будем ? К тому же, я не вижу, тестовых курлов круда для юзера

DaloreVacAtion replied:
Удалил

maximovd commented on tests/conftest.py:
```diff
@@ -0,0 +1,87 @@
+
+
+
+# @pytest.fixture(scope="session")
```
Если не нужно, удалить

DaloreVacAtion replied:
Удалил

maximovd commented on tests/services.py:
```diff
@@ -0,0 +1,52 @@
+from typing import Optional
```
Так, вот тут я вообще не понял, ты либо что-то перепутал и закинул сервисы в тесты случайно, либо у тебя тесты, которые ничего не тестируют вообще :)

DaloreVacAtion replied:
У нас на проекте просто такой подход был, что для каждого модуля свои сервисы лежали в папке services, поэтому и тут так сделал. Подправлю

maximovd commented on tests/test_api_v1_router.py:
```diff
@@ -0,0 +1,63 @@
+    )
+    not_existing_user = await get_user_by_id(user_id, session)
+    assert response is not None
+    assert response.status_code == 204
```
Коды лучше брать из `starlette` 
```python
from starlette import status
...
assert response.status_code == status.HTTP_204_NO_CONTENT

DaloreVacAtion replied:
Добавил

maximovd commented on tests/test_auth.py:
```diff
@@ -0,0 +1,74 @@
+from tests.services import create_user, clear_users
+
+
+# Как протестировать это?
```
Не очень понял твой вопрос, предлагаю обсудить

DaloreVacAtion replied:
До этого возникали проблемы, забыл убрать комментарий

Commit:
ПОдправлены комментарии, некоторые функции убраны в сервисы, подправлены аннотации типов

Commit:
добавлен structlog, pydantic-settings

maximovd commented on api/v1/crud.py:
```diff
@@ -0,0 +1,22 @@
+
+async def update_user(user_id: int, user_in: UserUpdate, session: AsyncSession):
+    update_data = user_in.dict(exclude_unset=True)
+    stmt = (
```
Что значит `stmt` ? У нас какое-то ограничение на количество символов в названии перменной :) ?

DaloreVacAtion replied:
Просто увидел, что так делают, поэтому тоже так сделал

maximovd commented on auth/manager.py:
```diff
@@ -0,0 +1,17 @@
+from auth.utils import get_user_db
+from config import SECRET_KEY
+
+SECRET = SECRET_KEY
```
А вот это зачем ? Почему просто не использовать `SECRET_KEY`

DaloreVacAtion replied:
Заменил

maximovd commented on auth/schemas.py:
```diff
@@ -0,0 +1,41 @@
+        default=None, title='Возраст пользователя', gt=0,
+    )
+    is_active: Optional[bool] = True
+    is_superuser: Optional[bool] = False
```
Тот же коммент про `Optional` в python3.10 принято писать: 
```python
is_superuser: bool | None 
```
К тому же, почему оно вообще может быть `None` ? Может лучше оставить строгое `bool`, потому что потом могут быть казусы по типу, юзер не активен и не неактивен, а активность у него `None`, а мы пытаемся что-то с ним сделать и все это падает в рантайме,

DaloreVacAtion replied:
ЗАменил

maximovd commented on questions.txt:
```diff
@@ -0,0 +1,3 @@
+1) Почему этот код "query = select(User).where(User.username == username)" возвращал None, хотя аналогичный, только
```
Нужно смотреть на sql запрос, который он формирует в базу, скорее всего разница там

DaloreVacAtion replied:
Нашёл решение

maximovd commented on questions.txt:
```diff
@@ -0,0 +1,3 @@
+1) Почему этот код "query = select(User).where(User.username == username)" возвращал None, хотя аналогичный, только
+через text(f'сырой запрос') возвращает корректного пользователя?
+2) При обновлении пользователя (api/v1/touter.py 40 строка), возвращает не обновленного пользователя, а ошибку, не могу понять, что не так
```
Надо глянуть на ошибку для начала

DaloreVacAtion replied:
Нашёл решение

maximovd commented on services/users.py:
```diff
@@ -0,0 +1,65 @@
+
+
+async def get_user_cookies(ac: AsyncClient, email: str, password: str) -> str | None:
+    auth = await ac.post('api/v1/auth/login', data={
```
Что выглядит очень костыльно, ты где это вообще взял ?

DaloreVacAtion replied:
Заменил

maximovd commented on services/users.py:
```diff
@@ -0,0 +1,65 @@
+        "username": email,
+        "password": password,
+    })
+    cookie = auth.cookies.get('IDP') if auth.status_code == 204 else None
```
Помнишь, что я тебе говорил про статусы ? `from starlette import status`

DaloreVacAtion replied:
Заменил

maximovd commented on tests/conftest.py:
```diff
@@ -0,0 +1,79 @@
+        yield conn
+
+
+# SETUP
```
Что это за коммент ? Какую смысловую нагрузку он несет ?

DaloreVacAtion replied:
Хотел пометить, что тут настройка для тестирования

maximovd commented on tests/test_auth.py:
```diff
@@ -0,0 +1,73 @@
+
+async def test_register_user_when_user_exist(async_client: AsyncClient):
+    create_data = {
+        'email': 'testmail@gmail.com',
```
Посмотри в сторону https://github.com/litestar-org/polyfactory Это аналог factoryboy для fastapi, позволяет использовать фабрики для генерации случайных данных

DaloreVacAtion replied:
Попробовал, отписался

Commit:
stmt заменён на statement, заменил secret на secret_key в мэнеджере пользователя, убрал Optional, заменил на | из python3.10, заменил числовые статусы на статусы из starlette

Commit:
stmt заменён на statement, заменил secret на secret_key в мэнеджере пользователя, убрал Optional, заменил на | из python3.10, заменил числовые статусы на статусы из starlette

Commit:
- Убран JWT Cookies, заменён на стандартный JWT Access Token
- Исправлен вывод пользователя в routers
- стандартный корявый конфиг с настройками заменён на pydantic-settings
- ПОдправлены тесты, а именно убран JWT Cookies

Commit:
- Вспомогательные функции для тестов перемещены в helpers

Commit:
- Добавил пропущенные импорты, попробовал фабрики для тестов

Commit:
Раскомментировал тесты, которые случайно закомментил

maximovd commented on api/v1/crud.py:
```diff
@@ -0,0 +1,22 @@
+
+
+async def delete_user(user: User, session: AsyncSession):
+    statement = text(f'delete from public.user where public.user.id = {user.id}')
```
Вроде же договорились через `where` делать ?

DaloreVacAtion replied:
Да, согласен, забыл убрать

maximovd commented on api/v1/crud.py:
```diff
@@ -0,0 +1,22 @@
+
+
+async def delete_user(user: User, session: AsyncSession):
+    statement = text(f'delete from public.user where public.user.id = {user.id}')
```
Вроде же договорились через `where` делать ?

DaloreVacAtion replied:
Да, согласен, забыл убрать

maximovd commented on api/v1/router.py:
```diff
@@ -0,0 +1,61 @@
+    """Обновить пользователя."""
+    user = await user_services.get_user_by_id(user_id, session)
+    await crud.update_user(user_id, user_in, session)
+    await session.refresh(user)
```
А вот этот `session.refresh()` нельзя убрать в сервис обновления пользователя ?

DaloreVacAtion replied:
Можно, да, согласен

DaloreVacAtion replied:
Убрал

maximovd commented on tests/conftest.py:
```diff
@@ -0,0 +1,78 @@
+
+metadata = Base.metadata
+
+# DATABASE
```
Опять же, не ясно, зачем тут этот коммент, по контексту и так ясно, что тут настройки базы данных

DaloreVacAtion replied:
Убрано

maximovd commented on tests/conftest.py:
```diff
@@ -0,0 +1,78 @@
+metadata = Base.metadata
+
+# DATABASE
+DATABASE_URL_TEST = settings.DB_TEST_URL
```
Вроде бы уже договорились, что так не делаем или тут другое ?

DaloreVacAtion replied:
Да, договорились, убрал

maximovd commented on tests/helpers.py:
```diff
@@ -0,0 +1,54 @@
+    return user
+
+
+async def create_user(
```
А зачем тебе нужен `create_user`, если ты добавил фабрику ? Можно же просто сделать User.build() и он тебе соберет пользователя

DaloreVacAtion replied:
Вот тут как раз и проблема, я пытался собрать пользователя, но в ответ приходит это
```
 polyfactory.exceptions.ConfigurationException: Model type User is not supported. To support it, register an appropriate base factory and subclass it for your factory.
```

maximovd replied:
Так тут же написано, что нужно сделать 
`To support it, register an appropriate base factory and subclass it for your factory.`

DaloreVacAtion replied:
Я поменял на Factory Boy, мне с ней проще было разобраться

Commit:
- Внёс правки по ревью, добавил создание пользователя через фабрику

Commit:
- Добавлена фабрика Factory Boy для тестов, произведеные тесты

maximovd commented on api/v1/crud.py:
```diff
@@ -0,0 +1,27 @@
+from auth.schemas import UserUpdate, UserRead
+
+
+async def delete_user(user: User, session: AsyncSession):
```
Сюда вопхнуть `-> None:`

maximovd commented on api/v1/router.py:
```diff
@@ -0,0 +1,60 @@
+async def login_user_token(
+    form_data: OAuth2PasswordRequestForm = Depends(),
+    session: AsyncSession = Depends(get_async_session),
+):
```
Тоже самое `-> dict[str, str]`. И вообще, почему не типизировано ?

maximovd commented on tests/conftest.py:
```diff
@@ -0,0 +1,82 @@
+
+
+@pytest.fixture(scope='session')
+async def get_db():
```
Везде проставь выходные типы

Commit:
- Добавлены выходные типы
